In [1]:
!dir


 Volume in drive C has no label.
 Volume Serial Number is 04FC-5BE8

 Directory of c:\D_Drive\DLproject\CreditCardFraudProject\notebooks

10/06/2025  11:59 PM    <DIR>          .
10/06/2025  11:59 PM    <DIR>          ..
10/06/2025  11:44 PM                 2 .gitkeep
10/08/2025  02:48 PM            10,792 preprocessing.ipynb
               2 File(s)         10,794 bytes
               2 Dir(s)  224,522,031,104 bytes free


In [2]:
import pandas as pd
import numpy as np
import sklearn
import matplotlib
import tensorflow as tf   # or torch, depending on what you use

print("✅ All libraries are working!")


✅ All libraries are working!


In [3]:
file_path = r"C:\D_Drive\DLproject\CreditCardFraudProject\data\CreditCardData.csv"

df = pd.read_csv(file_path)


print(df.shape)
print(df.info())
df.head()

(100000, 16)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 16 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   Transaction ID          100000 non-null  object 
 1   Date                    100000 non-null  object 
 2   Day of Week             100000 non-null  object 
 3   Time                    100000 non-null  int64  
 4   Type of Card            100000 non-null  object 
 5   Entry Mode              100000 non-null  object 
 6   Amount                  99994 non-null   object 
 7   Type of Transaction     100000 non-null  object 
 8   Merchant Group          99990 non-null   object 
 9   Country of Transaction  100000 non-null  object 
 10  Shipping Address        99995 non-null   object 
 11  Country of Residence    100000 non-null  object 
 12  Gender                  99996 non-null   object 
 13  Age                     100000 non-null  float64
 14  Bank    

Transaction ID       Date Day of Week  Time Type of Card Entry Mode Amount  \
0      #3577 209  14-Oct-20   Wednesday    19         Visa        Tap     £5   
1      #3039 221  14-Oct-20   Wednesday    17   MasterCard        PIN   £288   
2      #2694 780  14-Oct-20   Wednesday    14         Visa        Tap     £5   
3      #2640 960  13-Oct-20     Tuesday    14         Visa        Tap    £28   
4      #2771 031  13-Oct-20     Tuesday    23         Visa        CVC    £91   

  Type of Transaction Merchant Group Country of Transaction Shipping Address  \
0                 POS  Entertainment         United Kingdom   United Kingdom   
1                 POS       Services                    USA              USA   
2                 POS     Restaurant                  India            India   
3                 POS  Entertainment         United Kingdom            India   
4              Online    Electronics                    USA              USA   

  Country of Residence Gender   Age      Bank  Fraud  
0       United Kingdom      M  25.2       RBS      0  
1                  USA      F  49.6    Lloyds      0  
2                India      F  42.2  Barclays      0  
3       United Kingdom      F  51.0  Barclays      0  
4       United Kingdom      M  38.0   Halifax      1

In [4]:
df['Bank'].value_counts()



Bank
Barclays    29936
Monzo       10103
RBS         10046
Metro       10012
Barlcays     9993
Halifax      9988
Lloyds       9965
HSBC         9957
Name: count, dtype: int64

In [5]:
df['Type of Card'].value_counts()

Type of Card
Visa          53812
MasterCard    46188
Name: count, dtype: int64

In [6]:
df['Day of Week'].value_counts()

Day of Week
Wednesday    50177
Tuesday      49797
Friday          14
Thursday        12
Name: count, dtype: int64

In [7]:
df['Time'].value_counts()

Time
13    5810
16    5785
12    5754
23    5750
11    5748
7     5717
14    5686
15    5685
21    5684
8     5683
17    5665
22    5622
19    5622
10    5610
9     5570
18    5548
20    5522
2      536
3      531
6      515
0      509
5      489
1      484
4      474
24       1
Name: count, dtype: int64

In [8]:
df['Entry Mode'].value_counts()

Entry Mode
PIN    49976
CVC    33481
Tap    16543
Name: count, dtype: int64

In [9]:
df['Merchant Group'].value_counts()

Merchant Group
Children         10679
Restaurant       10078
Services         10017
Gaming           10011
Fashion           9987
Subscription      9977
Food              9847
Entertainment     9839
Products          9839
Electronics       9716
Name: count, dtype: int64

In [10]:
df['Shipping Address'].value_counts()

Shipping Address
United Kingdom    60905
USA               10042
Russia             9799
China              9698
India              9551
Name: count, dtype: int64

In [11]:
df['Country of Residence'].value_counts()

Country of Residence
United Kingdom    81859
Russia             4618
USA                4590
China              4529
India              4404
Name: count, dtype: int64

In [12]:
df['Type of Transaction'].value_counts()

Type of Transaction
Online    33481
ATM       33274
POS       33245
Name: count, dtype: int64

In [13]:
df['Gender'].value_counts()

Gender
M    50875
F    49121
Name: count, dtype: int64

In [14]:
df['Country of Transaction'].value_counts()

Country of Transaction
United Kingdom    71199
USA                7298
Russia             7257
China              7152
India              7094
Name: count, dtype: int64

In [15]:
df['Amount'].value_counts()

Amount
£17     2153
£23     2106
£16     2084
£28     2077
£15     2073
        ... 
£124     106
£145     105
£380     102
£375     100
£393      99
Name: count, Length: 396, dtype: int64

In [16]:
df['Fraud'].value_counts()

Fraud
0    92805
1     7195
Name: count, dtype: int64

Split Dataset by Fraud Status

In [17]:
fraud_df = df[df['Fraud'] == 1]
non_fraud_df = df[df['Fraud'] == 0]

We’ll take all 7,195 fraud rows and 7,805 random non-fraud rows to make a total of 15,000 rows.


In [46]:

non_fraud_sample = non_fraud_df.sample(n=7805, random_state=42)
newDf = pd.concat([fraud_df, non_fraud_sample])

Shuffle the Combined Dataset

In [47]:
 newDf= newDf.sample(frac=1, random_state=42).reset_index(drop=True)

In [ ]:
import os

# Create "data" folder if not exist
os.makedirs("data", exist_ok=True)

# Now save the file
newDf.to_csv("C:\D_Drive\DLproject\CreditCardFraudProject\data\CreditCardData_15000.csv", index=False)

print("Dataset saved successfully with shape:", newDf.shape)

Dataset saved successfully with shape: (14992, 16)


<>:7: SyntaxWarning: invalid escape sequence '\D'
<>:7: SyntaxWarning: invalid escape sequence '\D'
C:\Users\Vishva\AppData\Local\Temp\ipykernel_16344\2818062338.py:7: SyntaxWarning: invalid escape sequence '\D'
  newDf.to_csv("C:\D_Drive\DLproject\CreditCardFraudProject\data\CreditCardData_15000.csv", index=False)


In [ ]:
# Verify new distribution
newDf['Fraud'].value_counts()


Fraud
0    7805
1    7195
Name: count, dtype: int64

In [ ]:
newDf.head()

Transaction ID       Date Day of Week  Time Type of Card Entry Mode Amount  \
0      #2800 777  14-Oct-20   Wednesday    19         Visa        CVC    £21   
1      #3546 149  13-Oct-20     Tuesday     0         Visa        CVC     £6   
2      #3203 063  13-Oct-20     Tuesday     9   MasterCard        PIN     £7   
3      #2654 293  14-Oct-20   Wednesday    20         Visa        CVC    £17   
4      #3195 835  14-Oct-20   Wednesday    11         Visa        CVC    £27   

  Type of Transaction Merchant Group Country of Transaction Shipping Address  \
0              Online         Gaming         United Kingdom            China   
1              Online         Gaming                 Russia           Russia   
2                 ATM  Entertainment         United Kingdom   United Kingdom   
3              Online       Children         United Kingdom            China   
4              Online           Food         United Kingdom            India   

  Country of Residence Gender   Age      Bank  Fraud  
0       United Kingdom      M  46.1    Lloyds      0  
1       United Kingdom      M  52.0     Metro      1  
2       United Kingdom      M  43.9     Monzo      0  
3       United Kingdom      F  31.5  Barclays      1  
4       United Kingdom      F  33.8  Barlcays      1

In [ ]:
newDf.shape

(15000, 16)

In [ ]:
newDf.isnull().sum()

Transaction ID            0
Date                      0
Day of Week               0
Time                      0
Type of Card              0
Entry Mode                0
Amount                    3
Type of Transaction       0
Merchant Group            0
Country of Transaction    0
Shipping Address          0
Country of Residence      0
Gender                    1
Age                       0
Bank                      0
Fraud                     0
dtype: int64

In [ ]:
newDf = newDf.dropna()

In [ ]:
newDf.isnull().sum()

Transaction ID            0
Date                      0
Day of Week               0
Time                      0
Type of Card              0
Entry Mode                0
Amount                    0
Type of Transaction       0
Merchant Group            0
Country of Transaction    0
Shipping Address          0
Country of Residence      0
Gender                    0
Age                       0
Bank                      0
Fraud                     0
dtype: int64

In [ ]:
newDf.shape

(14996, 16)

In [ ]:
newDf['Bank'].value_counts()

Bank
Barclays    4605
Metro       1550
Barlcays    1527
Monzo       1520
RBS         1465
Lloyds      1453
Halifax     1440
HSBC        1436
Name: count, dtype: int64

In [ ]:
newDf['Type of Card'].value_counts()

Type of Card
Visa          8462
MasterCard    6534
Name: count, dtype: int64

In [ ]:
newDf['Day of Week'].value_counts()

Day of Week
Tuesday      7516
Wednesday    7476
Friday          2
Thursday        2
Name: count, dtype: int64

Remove tha Day of week = Friday / Thursday

In [ ]:
newDf = newDf[~newDf['Day of Week'].isin(['Friday', 'Thursday'])]

newDf = newDf.reset_index(drop=True)

In [ ]:
newDf['Day of Week'].value_counts()

Day of Week
Tuesday      7516
Wednesday    7476
Name: count, dtype: int64

In [ ]:
newDf['Date'].value_counts()

Date
13-Oct-20    7516
14-Oct-20    7476
Name: count, dtype: int64

In [ ]:
newDf['Time'].value_counts()

Time
13    800
22    728
7     705
8     685
15    677
10    676
20    675
18    672
16    665
11    664
17    663
23    662
9     657
12    651
21    633
14    626
19    616
2     536
3     531
6     515
0     509
5     488
1     483
4     474
24      1
Name: count, dtype: int64

In [49]:
# Only numeric features
numeric_cols = newDf.select_dtypes(include=['int64', 'float64']).columns.drop('Fraud')

# Correlation with target
correlation = newDf[numeric_cols].corrwith(newDf['Fraud']).sort_values(key=abs, ascending=False)

# Convert to percentage (absolute value)
correlation_percent = (correlation.abs() / correlation.abs().sum()) * 100

print(correlation_percent)


Time    98.51103
Age      1.48897
dtype: float64


In [50]:
from sklearn.preprocessing import LabelEncoder
from sklearn.feature_selection import chi2

# Encode categorical features
cat_cols = newDf.select_dtypes(include=['object']).columns
le = LabelEncoder()

X_cat = newDf[cat_cols].apply(lambda col: le.fit_transform(col))
y = newDf['Fraud']

# Chi-square test
chi_scores, p_values = chi2(X_cat, y)

# Convert to percentages
chi_percent = (chi_scores / chi_scores.sum()) * 100

for col, score in zip(cat_cols, chi_percent):
    print(f"{col}: {score:.2f}%")


Transaction ID: 15.27%
Date: 0.07%
Day of Week: 0.01%
Type of Card: 0.26%
Entry Mode: 0.01%
Amount: 17.60%
Type of Transaction: 0.32%
Merchant Group: 4.60%
Country of Transaction: 26.58%
Shipping Address: 33.27%
Country of Residence: 1.53%
Gender: 0.42%
Bank: 0.04%


In [53]:
import pandas as pd

# Filter only fraud transactions
fraud_df = newDf[newDf['Fraud'] == 1]

# Loop through all columns
for col in newDf.columns:
    print(f"--- {col} ---")
    
    if newDf[col].dtype == 'object':  # Categorical column
        counts = fraud_df[col].value_counts()
        percent = (counts / counts.sum()) * 100
        summary = pd.DataFrame({
            'Count (Fraud=1)': counts,
            'Percentage (%)': percent.round(2)
        })
    else:  # Numeric column
        summary = pd.DataFrame({
            'Mean (Fraud=1)': [fraud_df[col].mean()],
            'Median (Fraud=1)': [fraud_df[col].median()],
            'Min (Fraud=1)': [fraud_df[col].min()],
            'Max (Fraud=1)': [fraud_df[col].max()]
        })
    
    print(summary)
    print()


--- Transaction ID ---
                Count (Fraud=1)  Percentage (%)
Transaction ID                                 
#3462 932                     2            0.03
#3090 262                     2            0.03
#3198 303                     2            0.03
#3293 336                     2            0.03
#2547 017                     2            0.03
...                         ...             ...
#2708 292                     1            0.01
#2994 784                     1            0.01
#3414 350                     1            0.01
#3538 621                     1            0.01
#3195 835                     1            0.01

[7171 rows x 2 columns]

--- Date ---
           Count (Fraud=1)  Percentage (%)
Date                                      
13-Oct-20             3734           51.90
14-Oct-20             3459           48.08
15-Oct-20                1            0.01
16-Oct-20                1            0.01

--- Day of Week ---
             Count (Fraud=1)  Perce